### PR TITLE
MM-52898: WS fix

### PR DESCRIPTION
**Pull request by agnivade**
https://mattermost.atlassian.net/browse/MM-52898

```release-note
NONE
```

### CODE REVIEW

File: server/channels/app/server.go [hunk 1]
```diff
@@ -1152,7 +1152,19 @@ func (a *App) OriginChecker() func(*http.Request) bool {
...
+	return func(r *http.Request) bool {
+		origin := r.Header["Origin"]
+		if len(origin) == 0 {
+			return true
+		}
+		u, err := url.Parse(origin[0])
+		if err != nil {
+			return false
+		}
+		return strings.EqualFold(u.Host, r.Host) && strings.EqualFold(u.Scheme, r.URL.Scheme)
```
Comment (jupenur) [inline server/channels/app/server.go:1166]: If SSL termination is done in a reverse proxy (nginx), will `r.URL.Scheme` still be HTTPS here?

Comment (agnivade) [reply server/channels/app/server.go:1166]: Ah yeah good point. What do you suggest?

Comment (jupenur) [reply server/channels/app/server.go:1166]: Use the site URL I guess? This is the point why this issue was so tricky to begin with though. I don't know if we'll run into issues with misconfigured site URLs.

Comment (agnivade) [reply server/channels/app/server.go:1166]: Interesting though, it looks like we are completely ignoring `WebsocketURL` in the code. It's not being used anywhere AFAI can see.

Comment (jupenur) [reply server/channels/app/server.go:1166]: The websocket URL is used by the client, but the site URL is what we care about here.

Comment (agnivade) [reply server/channels/app/server.go]: PTAL

File: server/channels/app/server.go [hunk 2]
```diff
@@ -1152,7 +1152,28 @@ func (a *App) OriginChecker() func(*http.Request) bool {
...
+		u, err := url.Parse(origin[0])
+		if err != nil {
+			return false
+		}
+
+		// To maintain the case where siteURL is not set.
+		if *a.Config().ServiceSettings.SiteURL == "" {
+			return strings.EqualFold(u.Host, r.Host)
+		}
+		siteURL, err := url.Parse(*a.Config().ServiceSettings.SiteURL)
```
Comment (jupenur) [inline server/channels/app/server.go:1171]: Should we cover misconfigurations like the siteURL being set to `https:\\example.com\` -- which would presumably work for all cases that don't invoke `net/url.Parse`, but with `net/url` results in an opaque URL.

0/5

Comment (agnivade) [reply server/channels/app/server.go:1171]: Hmm .. but then how would you actually parse out the scheme and host in that case? Seems a bit of edge case to me to try to handle that.

Comment (jupenur) [reply server/channels/app/server.go:1171]: Yeah probably doesn't make sense trying to parse bad urls like that, but we could try to detect them and bail out:

```go
if siteURL.Host == "" || siteURL.Opaque != "" {
  return false
}
```

Comment (agnivade) [reply server/channels/app/server.go:1171]: I am thinking of keep this as is for now to keep things consistent with the rest of the codebase. I don't think we do special handling of what happens if a siteURL is opaque. Either it parses or not. Maybe if the need arises, we handle that all over the codebase in a separate PR?

Comment (agnivade) [reply server/channels/app/server.go:1174]: done

**Review by jupenur — APPROVED** [review]
LGTM

File: server/channels/app/server.go [hunk 3]
```diff
@@ -1152,7 +1152,31 @@ func (a *App) OriginChecker() func(*http.Request) bool {
...
+
+	// Overriding the default origin checker
+	return func(r *http.Request) bool {
+		origin := r.Header["Origin"]
+		if len(origin) == 0 {
+			return true
+		}
+		if origin[0] == "null" {
+			return false
+		}
```
Comment (streamer45) [inline server/channels/app/server.go:1164]: What is this case covering exactly? It's not super clear from this side of things.

Comment (agnivade) [reply server/channels/app/server.go:1164]: It got lost in the previous discussion https://github.com/mattermost/mattermost-server/pull/23489#discussion_r1204051836

Comment (jupenur) [reply server/channels/app/server.go:1164]: For additional context, MDN does a good job of explaining null origins: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin#description

Comment (streamer45) [reply server/channels/app/server.go:1164]: Thanks both :+1:

**Review by streamer45 — APPROVED** [review]

**Comment by agnivade** [comment]
Only 5 failures, and doesn't appear to be related. Merging.

**Pull request merged**